### PR TITLE
Implement OmnitelK Exporter

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Implement OmnitelK Exporter

Exporter glues Client and Encoder together and
implements the top-level logic according to the design
https://docs.google.com/document/d/1eGa9e08rfqJWvWHRcP-8u4y1dj1FqWpOGQS86Le1Om8/edit#heading=h.wdhqyf7yh95x

Testing done: make test

TODO:
Need to add more tests that verify various edge
conditions (e.g. timeouts, connection failures, etc).

### CODE REVIEW

File: exporter/omnitelk/exporter.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	e.client.Send(record, originalSpans, shard.origin)
+}
+
+func (e *Exporter) resendSpans(spans []*jaeger.Span) {
+	// Push the failed spans to the retry queue to be picked up by processRetryQueue later.
+	select {
+	case e.spansToRetry <- spans:
+	default:
+		// No more room in spansToRetry queue. Drop the failed spans.
+		e.encoder.hooks.OnDropSpans(int64(len(spans)))
```
Comment (owais) [inline exporter/omnitelk/exporter.go]: IIRC historically we used to drop from the front of the queue while adding new data to the back so in a situation where we drop for a long period of time, we'll at least try to keep pushing latest records instead of the oldest ones.

File: exporter/omnitelk/client_unary.go [hunk 2]
```diff
@@ -145,18 +145,12 @@ func (c *ClientUnary) processSendRequests() {
...
-			}
+			c.sendRequest(request)
 		}
 	}
 }
 
-func (c *ClientUnary) sendRequest(
-	pr requestToSend,
-) bool {
+func (c *ClientUnary) sendRequest(pr requestToSend) {
```
Comment (pjanotti) [inline exporter/omnitelk/client_unary.go]: ah the issue that I mentioned in the other PR is fixed here

File: exporter/omnitelk/encoder.go [hunk 3]
```diff
@@ -94,32 +94,28 @@ type encoder struct {
...
+	// Minimum and default value is 1.
 	NumWorkers uint
 
 	// MaxRecordSize is the maximum size of one encoded uncompressed record in bytes.
 	// When the record reaches this size, it gets sent to the server.
 	MaxRecordSize uint
 
 	// BatchFlushInterval defines how often to flush batches of spans if they don't
-	// reach the MaxRecordSize. The default value is defaultBatchFlushInterval.
+	// reach the MaxRecordSize. The default value is defBatchFlushInterval.
```
Comment (pjanotti) [inline exporter/omnitelk/encoder.go]: personal preference but I liked the old name...

File: exporter/omnitelk/exporter.go (hunk 1 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/exporter.go]: To implement that I will need implement custom queues, I don't think I can easily achieve that using regular channel. For the sake of simplicity I would like to keep this and if we see drops happening in from there we can re-think. Realistically we should not hit this. Retries happen when config mismatch is detected and re-encoding is needed. It should be very rare and retry queue should be always empty and enough to process all that is returned for processing. I don't expect any hits on this "drop", if it happens we can decide how to address it best.

File: exporter/omnitelk/client_unary.go (hunk 2 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/client_unary.go]: yes :-)

File: exporter/omnitelk/factory.go [hunk 4]
```diff
@@ -26,7 +28,13 @@ const (
 	typeStr = "omnitelk"
 
 	// Default values for config options.
-	defSendConcurrency = 20
+	defSendConcurrency          = 20
+	defStreamReopenPeriod       = 30 * time.Second
+	defStreamReopenRequestCount = 1000
+	defNumWorkers               = 1
+	defMaxRecordSize            = 100000
+	defBatchFlushInterval       = time.Second * 5
```
Comment (ccaraman) [inline exporter/omnitelk/factory.go]: super super nit: this definitino is in a separate order from `defStreamReopenPeriod`

File: exporter/omnitelk/exporter.go (hunk 1 above)
Comment (owais) [reply exporter/omnitelk/exporter.go]: Fine for now but I think we should prepare for the rare case by at least discussing and agreeing what behavior we want in the case. May be add an internal ticket to revisit this and decide what to do sometime in near future? 

BTW historically we've used this: https://github.com/jaegertracing/jaeger/tree/master/pkg/queue

Comment (tigrannajaryan) [reply exporter/omnitelk/exporter.go]: I am invoking my YAGNI rights :-)

I am going to add telemetry to warn us if this happens.

File: exporter/omnitelk/factory.go (hunk 4 above)
Comment (tigrannajaryan) [reply exporter/omnitelk/factory.go]: Fixed.
